arangod/IResearch/GeoAnalyzer.cpp

```diff
@@ -290,93 +362,144 @@ void GeoAnalyzer::reset(std::vector<std::string>&& terms) noexcept {
+  std::vector<std::string> geoTerms;
+  auto const type = _shape.type();
+  if (_type == Type::CENTROID || type == geo::ShapeContainer::Type::S2_POINT) {
+    geoTerms = _indexer.GetIndexTerms(_shape.centroid(), {});
+  } else if (legacyCovering) {
+    geoTerms = _indexer.GetIndexTerms(*_shape.region(), {});
+  } else {
+    auto cellIds = _shape.covering(_coverer);
+    // TODO(MBkkt) S2_MULTIPOINT can be handled via GetIndexTerms(point)
+    if (type == geo::ShapeContainer::Type::S2_MULTIPOINT ||
+        type == geo::ShapeContainer::Type::S2_MULTIPOLYLINE) {
+      _coverer.CanonicalizeCovering(&cellIds);
+    } else {
+      TRI_ASSERT(_coverer.IsCanonical(cellIds));
+    }
+    geoTerms = _indexer.GetIndexTermsForCanonicalCovering(
+        S2CellUnion::FromVerbatim(std::move(cellIds)), {});
+  }
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+  std::vector<std::string_view> terms{geoTerms.begin(), geoTerms.end()};
+  std::sort(terms.begin(), terms.end());
+  TRI_ASSERT(std::unique(terms.begin(), terms.end()) == terms.end());
+#endif
+  GeoAnalyzer::reset(std::move(geoTerms));
+  return true;
 }
 
-/*static*/ VPackSlice GeoJSONAnalyzer::store(
-    irs::token_stream const* ctx, VPackSlice slice,
-    velocypack::Buffer<uint8_t>& buf) noexcept {
-  TRI_ASSERT(ctx);
-  auto const* impl = basics::downCast<GeoJSONAnalyzer>(ctx);
-  if (Type::CENTROID == impl->shapeType()) {
-    TRI_ASSERT(!impl->_shape.empty());
-    S2LatLng const centroid(impl->_shape.centroid());
+bool GeoVPackAnalyzer::normalize(std::string_view args, std::string& out) {
+  return normalizeImpl<GeoVPackAnalyzer>(args, out);
+}
 
-    VPackBuilder array(buf);
-    toVelocyPack(array, centroid);
+irs::analysis::analyzer::ptr GeoVPackAnalyzer::make(std::string_view args) {
+  return makeImpl<GeoVPackAnalyzer>(args);
+}
 
-    return array.slice();
+irs::bytes_view GeoVPackAnalyzer::store(irs::token_stream* ctx,
+                                        velocypack::Slice slice) {
+  TRI_ASSERT(ctx != nullptr);
+  auto& impl = basics::downCast<GeoVPackAnalyzer>(*ctx);
+  if (impl._type == Type::CENTROID) {
+    TRI_ASSERT(!impl._shape.empty());
+    S2LatLng const centroid{impl._shape.centroid()};
+    impl._builder.clear();
+    toVelocyPack(impl._builder, centroid);
+    slice = impl._builder.slice();
   }
-
-  return slice;
+  return ref<irs::byte_type>(slice);
 }
 
-GeoJSONAnalyzer::GeoJSONAnalyzer(Options const& opts)
-    : GeoAnalyzer(irs::type<GeoJSONAnalyzer>::get()),
-      _indexer(S2Options(opts.options)),
-      _type(opts.type) {}
+GeoVPackAnalyzer::GeoVPackAnalyzer(Options const& options)
+    : GeoJsonAnalyzerBase{irs::type<GeoVPackAnalyzer>::get(), options},
+      _legacy{options.legacy} {}
 
-void GeoJSONAnalyzer::prepare(S2RegionTermIndexer::Options& opts) const {
-  opts = _indexer.options();
-  opts.set_index_contains_points_only(_type != Type::SHAPE);
+bool GeoVPackAnalyzer::reset(std::string_view value) {
+  // TODO(MBkkt) legacyCovering should be equal _legacy
```

Comment:
Do we really need to preserve "legacy covering" for non-legacy vpack analyzer?

Reply:
I think we don't need, @gnusi think we need